File: workflow/installed/check_installed.jl
# Installed-package check for Julia (claim
# treescape-installed-packages-reproduce-gallery).
#
#   julia workflow/installed/check_installed.jl <repo root> <report.json>
#
# Installs Treescape.jl into a fresh temporary environment, asserts the
# connector library comes from the release ARTIFACT (not a development
# build or an override), runs the parity cases, and compares every case
# that names a committed gallery file byte for byte.

using Pkg
repo, report = ARGS
env = mktempdir()
Pkg.activate(env)
Pkg.develop(path = joinpath(repo, "packages", "Treescape.jl"))
Pkg.add("TOML")
using Treescape, TOML

haskey(ENV, "TREESCAPE_JL_LIB") && error("TREESCAPE_JL_LIB is set; the check must use the artifact")
lib = Treescape.library_path()
occursin(joinpath("artifacts", ""), lib) || error("connector library not from an artifact: $lib")

out = mktempdir()
runner = joinpath(repo, "packages", "Treescape.jl", "test", "parity_runner.jl")
cases = joinpath(repo, "tests", "fixtures", "parity", "cases.toml")
run(`$(Base.julia_cmd()) --project=$env $runner $cases $repo $out`)

results = Dict{String,Bool}()
for case in TOML.parsefile(cases)["cases"]
    haskey(case, "expect_file") || continue
    got = read(joinpath(out, case["name"] * ".svg"), String)
    expected = read(joinpath(repo, case["expect_file"]), String)
    results[case["expect_file"]] = got == expected
    got == expected || replace(expected, "\r\n" => "\n") != got ||
        println("$(case["expect_file"]): differs in line endings only (is the checkout converting them? see .gitattributes)")
end
mismatches = sort([k for (k, ok) in results if !ok])
open(report, "w") do io
    TOML.print(io, Dict("claim" => "treescape-installed-packages-reproduce-gallery", "platform" => string(Sys.MACHINE),
        "julia" => string(VERSION), "library" => lib, "checked" => length(results), "mismatches" => mismatches))
end
println("checked $(length(results)) gallery files; mismatches: $mismatches")
isempty(mismatches) || exit(1)

File: workflow/installed/check_installed.py
"""Installed-package check (claim treescape-installed-packages-reproduce-gallery).

    python workflow/installed/check_installed.py <repo root> <report.json>

Run from a directory OUTSIDE the repository, in a virtual environment into
which only the built wheels were installed. Asserts that ``treescape`` and
``treescape_connector`` are imported from the environment (never from the
repository's ``packages/`` sources), that the connector wheel carries both
license files, and that every gallery and README SVG
(``scripts/regen_assets.py``) is reproduced byte for byte.
"""

from __future__ import annotations

import importlib.metadata
import importlib.util
import json
import platform
import sys
from pathlib import Path


def main(repo: Path, report: Path) -> int:
    import treescape
    import treescape_connector
    import treescape_reference

    for module in (treescape, treescape_connector, treescape_reference):
        where = Path(module.__file__).resolve()
        assert repo.resolve() not in where.parents, f"{module.__name__} imported from the repository ({where})"

    files = {Path(f).name for f in importlib.metadata.files("treescape_connector") or []}
    missing = {"LICENSE", "LICENSE.DejaVu.txt"} - files
    assert not missing, f"connector wheel lacks {sorted(missing)}"
    font = Path(treescape_reference.__file__).parent / "fonts" / "DejaVuSans.ttf"
    assert font.is_file(), f"treescape-reference wheel lacks its font ({font})"

    spec = importlib.util.spec_from_file_location("regen_assets", repo / "scripts" / "regen_assets.py")
    regen = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(regen)
    results = {}
    for name, build in regen.GALLERY:
        results[name] = build().to_svg() == (regen.GALLERY_DIR / name).read_text(encoding="utf-8")
    results["primates.svg"] = regen._marketing().to_svg() == regen.TARGET.read_text(encoding="utf-8")

    summary = {
        "claim": "treescape-installed-packages-reproduce-gallery",
        "platform": f"{platform.system()}-{platform.machine()}",
        "python": platform.python_version(),
        "treescape": importlib.metadata.version("treescape"),
        "treescape_connector": importlib.metadata.version("treescape-connector"),
        "files": results,
        "mismatches": sorted(k for k, ok in results.items() if not ok),
    }
    report.write_text(json.dumps(summary, indent=2, sort_keys=True))
    print(json.dumps(summary, indent=2, sort_keys=True))
    return 1 if summary["mismatches"] else 0


if __name__ == "__main__":
    raise SystemExit(main(Path(sys.argv[1]), Path(sys.argv[2])))
